crates/ruff_linter/src/rules/ruff/rules/pytest_raises_ambiguous_pattern.rs

```diff
@@ -0,0 +1,161 @@
+/// `re.escape()`.
+///
+/// ## Example
+///
+/// ```python
+/// import pytest
+///
+///
+/// with pytest.raises(Exception, match="A full sentence."):
+///     do_thing_that_raises()
+/// ```
+///
+/// Use instead:
+///
+/// ```python
+/// import pytest
+///
+///
+/// with pytest.raises(Exception, match=r"A full sentence."):
+///     do_thing_that_raises()
+/// ```
+///
+/// Alternatively:
+///
+/// ```python
+/// import pytest
+/// import re
+///
+///
+/// with pytest.raises(Exception, match=re.escape("A full sentence.")):
+///     do_thing_that_raises()
+/// ```
+///
+/// or:
+///
+/// ```python
+/// import pytest
+/// import re
+///
+///
+/// with pytest.raises(Exception, "A full sentence\\."):
+///     do_thing_that_raises()
+/// ```
+///
+/// ## References
+/// - [Python documentation: `re.escape`](https://docs.python.org/3/library/re.html#re.escape)
+/// - [`pytest` documentation: `pytest.raises`](https://docs.pytest.org/en/latest/reference/reference.html#pytest-raises)
+#[derive(ViolationMetadata)]
+pub(crate) struct PytestRaisesAmbiguousPattern;
+
+impl Violation for PytestRaisesAmbiguousPattern {
+    #[derive_message_formats]
+    fn message(&self) -> String {
+        "Pattern passed to `match=` contains metacharacters but is neither escaped nor raw"
+            .to_string()
+    }
+
+    fn fix_title(&self) -> Option<String> {
+        Some("Use a raw string or `re.escape()` to make the intention explicit".to_string())
+    }
+}
+
+/// RUF043
+pub(crate) fn pytest_raises_ambiguous_pattern(checker: &mut Checker, call: &ExprCall) {
+    if !is_pytest_raises(&call.func, checker.semantic()) {
+        return;
+    }
+
+    let Some(Keyword { value, .. }) = call.arguments.find_keyword("match") else {
+        return;
+    };
+
+    let Some(string) = value.as_string_literal_expr() else {
+        return;
+    };
+
+    let any_part_is_raw = string.value.iter().any(|part| part.flags.prefix().is_raw());
+
+    if any_part_is_raw || !string_has_unescaped_metacharacters(&string.value) {
+        return;
+    }
+
+    let diagnostic = Diagnostic::new(PytestRaisesAmbiguousPattern, string.range);
+
+    checker.diagnostics.push(diagnostic);
+}
+
+fn string_has_unescaped_metacharacters(value: &StringLiteralValue) -> bool {
+    let mut escaped = false;
+    let mut last = '\x00';
+
+    for (current, next) in value.chars().tuple_windows() {
+        last = next;
+
+        if escaped {
+            if escaped_char_is_regex_metasequence(current) {
+                return true;
+            }
+
+            escaped = false;
+            continue;
+        }
+
+        if current == '\\' {
+            escaped = true;
+            continue;
+        }
+
+        if char_is_regex_metacharacter(current) {
+            return true;
+        }
+    }
+
+    if escaped {
+        escaped_char_is_regex_metasequence(last)
+    } else {
+        char_is_regex_metacharacter(last)
+    }
```

Comment:
It's not clear to me why this can't be handled in the for-loop directly. What's special about the last character?

Reply:
it's because we iterate through the `chars` in overlapping pairs using `tuple_windows()`. The final tuple yielded by `tuple_windows()` will be `(<penultimate character>, <final character>)`, i.e., the `current` variable in the `for` loop will never hold the value of the final character of the string